Riot/AppDelegate.m

```diff
@@ -957,71 +931,266 @@ - (void)application:(UIApplication*)application didReceiveRemoteNotification:(NS
                 {
-                    remoteNotificationRoomId = roomId;
-                    
-                    NSLog(@"[AppDelegate] didReceiveRemoteNotification: open the roomViewController %@", roomId);
-                    
-                    [self showRoom:roomId andEventId:nil withMatrixSession:dedicatedAccount.mxSession];
-                }
-                else
-                {
-                    NSLog(@"[AppDelegate] didReceiveRemoteNotification: busy");
-                }
-            }
-            else if (!_completionHandler && (state == UIApplicationStateBackground))
-            {
-                _completionHandler = completionHandler;
-                
-                NSLog(@"[AppDelegate] didReceiveRemoteNotification: starts a background sync");
-                
-                [dedicatedAccount backgroundSync:20000 success:^{
-                    NSLog(@"[AppDelegate] didReceiveRemoteNotification: the background sync succeeds");
-                    
-                    if (_completionHandler)
-                    {
-                        _completionHandler(UIBackgroundFetchResultNewData);
-                        _completionHandler = nil;
-                    }
-                } failure:^(NSError *error) {
-                    NSLog(@"[AppDelegate] didReceiveRemoteNotification: the background sync fails");
-                    
-                    if (_completionHandler)
+                    if (action.actionType == MXPushRuleActionTypeSetTweak)
                     {
-                        _completionHandler(UIBackgroundFetchResultNoData);
-                        _completionHandler = nil;
+                        if ([action.parameters[@"set_tweak"] isEqualToString:@"sound"])
+                        {
+                            NSString *soundName = action.parameters[@"value"];
+                            if ([soundName isEqualToString:@"default"])
+                                soundName = UILocalNotificationDefaultSoundName;
+                            
+                            eventNotification.soundName = soundName;
+                        }     
                     }
-                }];
+                }
                 
-                // wait that the background sync is done
-                return;
+                [[UIApplication sharedApplication] scheduleLocalNotification:eventNotification];
             }
         }
+        
+        dispatch_block_t completionBlock = [weakSelf createPushNotificationHandlingStateCleaningBlock];
+        weakSelf.pushNotificationHandlingCompletionBlock = completionBlock;
+        
+        // This delay will help us to process push events which we receive earlier than push notifications associated with them
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+            completionBlock();
+        });
+    };
+    
```

Comment:
@giomfo We need to decide which value will be appropriate here and i think it'll be better to assign this value to a constant